packages/generator-superset/generators/plugin-chart/templates/src/MyChart.erb

```diff
@@ -16,25 +17,122 @@
+
+    // Theming variables are provided for your use via a ThemeProvider
+    // imported from @superset-ui/style. For variables available, please visit
+    // https://github.com/apache-superset/superset-ui/blob/master/packages/superset-ui-style/src/index.ts
+
+    const Wrapper = styled.div`
+      background-color: ${({ theme }) => theme.colors.secondary.light2};
+      padding: ${({ theme }) => theme.gridUnit * 4}px;
+      border-radius: ${({ theme }) => theme.gridUnit * 2}px;
+      height: ${height};
+      width: ${width};
+      overflow-y: scroll;
+
+      h3 {
+        /* You can use your props to control CSS! */
+        font-size: ${({ theme }) => theme.typography.sizes[this.props.headerFontSize]};
+        font-weight: ${({ theme }) =>
+          theme.typography.weights[this.props.boldText ? 'bold' : 'normal']};
+      }
+    `;
+
+    console.log('Plugin props', this.props);
+
     return (
-      <div style={{ backgroundColor: '#ffe459', padding: 16, borderRadius: 8, height, width }}>
-        <h3>Hello!</h3>
-        <pre>
-          {JSON.stringify(this.props, null, 2)}
-        </pre>
-      </div>
+      <Wrapper ref={this.rootElem}>
+        <h3>{this.props.headerText}</h3>
+        <pre>{JSON.stringify(data, null, 2)}</pre>
+      </Wrapper>
     );
   }
-}
+}<% } %>
+<%if (componentType == 'function') { %>export default function <%= packageLabel %>(props: <%= packageLabel %>Props) {
+  // height and width are the height and width of the DOM element as it exists in the dashboard.
+  // There is also a `data` prop, which is, of course, your DATA 🎉
+  const { data, height, width } = props;
+
+  const rootElem = createRef<HTMLDivElement>();
+
+  // Often, you just want to get a hold of the DOM and go nuts.
+  // Here, you can do that with createRef, and the useEffect hook.
+  useEffect(() => {
+    const root = rootElem.current as HTMLElement;
+    console.log('Plugin element', root);
+  });
+
+  console.log('Plugin props', props);
+
+
+  // The following Wrapper is a <div> element, which has been styled using Emotion
+  // For docs, visit https://emotion.sh/docs/styled
+
+  // Theming variables are provided for your use via a ThemeProvider
+  // imported from @superset-ui/style. For variables available, please visit
+  // https://github.com/apache-superset/superset-ui/blob/master/packages/superset-ui-style/src/index.ts
+
+  const Wrapper = styled.div`
+    background-color: ${({ theme }) => theme.colors.secondary.light2};
+    padding: ${({ theme }) => theme.gridUnit * 4}px;
+    border-radius: ${({ theme }) => theme.gridUnit * 2}px;
+    height: ${height};
+    width: ${width};
+    overflow-y: scroll;
+
+    h3 {
+      /* You can use your props to control CSS! */
+      font-size: ${({ theme }) => theme.typography.sizes[props.headerFontSize]};
+      font-weight: ${({ theme }) => theme.typography.weights[props.boldText ? 'bold' : 'normal']};
+    }
+  `;
+
+  return (
+    <Wrapper ref={rootElem}>
```

Comment:
1. Does the `Wrapper` need `rootElem`?
2. It should probably be placed outside of the functional component otherwise a new styled component is created every time the chart component renders. Not sure about the performance impact of this, but that doesn't sound optimal.

I also like a convention I saw elsewhere which names styled divs `Styles`:

```jsx
const Styles = styled.div`
  h3 {
    /* You can use your props to control CSS! */
    font-size: ${({ theme, headerFontSize }) => theme.typography.sizes[headerFontSize]};
    font-weight: ${({ theme, boldText }) => theme.typography.weights[boldText ? 'bold' : 'normal']};
  }
`;

function MyComponent(props: { headerFontSize: string; boldText: boolean; }) {

  return (
    <Styles {...props}>
      ...
    </Styles>
  );
}
```

What do you think?

Reply:
Perfectly OK with the `Styled` name. It's more explicit. `Wrapper` felt a little dirty, since in a way, _everything's_ a wrapper.

As for whether the plugin needs the `ref`, that's just part of the example scaffolding so people can more easily grab a hold of the DOM element and go crazy on it with D3 (or whatever).

Now, as for _where_ to put that ref in the JSX, I'm not quite sure... I can _certainly_ take it up a layer, as you suggest, like so:
```
return (
  <div ref={rootElem}
    <Style>
      ...more stuff
    </Style>
  </div>
)
```
This would address your styled component instantiation concern. _However_, I'm not sure that's actually a problem... When I put two of these on a dash, they share the same class, which makes me think it's safe enough. Am I mistaken?

<img width="1024" alt="Pasted_Image_6_30_20__9_43_PM" src="https://user-images.githubusercontent.com/812905/86203949-248db800-bb1b-11ea-8bc2-48308d1dd03f.png">

<img width="1011" alt="Pasted_Image_6_30_20__9_46_PM" src="https://user-images.githubusercontent.com/812905/86203987-3cfdd280-bb1b-11ea-8996-3cf84a330c2b.png">

Reply:
ok, n/m, I think I understand where you're coming from... I'll externalize the Styled component.

Reply:
Yeah, emotion probably has proactive caching that makes this a non-issue, but externalizing it feels like a better practice because you can at least be sure you don't create a new styled component every time the chart re-renders---even though they may point to the same cache.

It will also help you 1) limit which props are available to styles 2) make it easier to extract it to another file when it becomes too large.